akka-http-core/src/main/scala/akka/http/impl/engine/client/ProxyGraphStage.scala

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.client
+
+import akka.event.LoggingAdapter
+import akka.http.impl.engine.parsing.ParserOutput.{ NeedMoreData, ResponseStart }
+import akka.http.impl.engine.parsing.{ HttpHeaderParser, HttpResponseParser }
+import akka.http.scaladsl.model.{ HttpMethods, StatusCodes }
+import akka.http.scaladsl.settings.ClientConnectionSettings
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+import akka.stream.{ Attributes, BidiShape, Inlet, Outlet }
+import akka.util.ByteString
+
+object ProxyGraphStage {
+  sealed trait State
+  // Entry state
+  case object Starting extends State
+
+  // State after CONNECT messages has been sent to Proxy and before Proxy responded back
+  case object Connecting extends State
+
+  // State after Proxy responded  back
+  case object Connected extends State
+}
+
+class ProxyGraphStage(targetHostName: String, targetPort: Int, settings: ClientConnectionSettings, log: LoggingAdapter)
+  extends GraphStage[BidiShape[ByteString, ByteString, ByteString, ByteString]] {
+
+  import ProxyGraphStage._
+
+  val bytesIn: Inlet[ByteString] = Inlet("OutgoingTCP.in")
+  val bytesOut: Outlet[ByteString] = Outlet("OutgoingTCP.out")
+
+  val sslIn: Inlet[ByteString] = Inlet("OutgoingSSL.in")
+  val sslOut: Outlet[ByteString] = Outlet("OutgoingSSL.out")
+
+  override def shape: BidiShape[ByteString, ByteString, ByteString, ByteString] = BidiShape.apply(sslIn, bytesOut, bytesIn, sslOut)
+
+  private val connectMsg = ByteString(s"CONNECT ${targetHostName}:${targetPort} HTTP/1.1\r\nHost: ${targetHostName}\r\n\r\n")
+
+  @scala.throws[Exception](classOf[Exception])
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var state: State = Starting
+
+    val parser = new HttpResponseParser(settings.parserSettings, HttpHeaderParser(settings.parserSettings, log)()) {
+      override def handleInformationalResponses = false
+    }
+    parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.CONNECT, None))
+
+    setHandler(sslIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting ⇒
+            throw new IllegalStateException("inlet OutgoingSSL.in unexpectedly pushed in Starting state")
+          case Connecting ⇒
+            throw new IllegalStateException("inlet OutgoingSSL.in unexpectedly pushed in Connecting state")
+          case Connected ⇒
+            push(bytesOut, grab(sslIn))
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
```

Comment:
I've added custom `onDownstreamFinish` and `onUpstreamFinish` to all inlets/outlets to be fully transparent for errors. Thanks to that in case Proxy is down we're getting `akka.stream.StreamTcpException: Tcp command [Connect(localhost:8888,None,List(),Some(10 seconds),true)] failed` instead of cryptic `akka.http.impl.engine.client.OutgoingConnectionBlueprint$UnexpectedConnectionClosureException: The http server closed the connection unexpectedly before delivering responses for 1 outstanding requests`. 

@jrudolph  Here I just want to ensure that it's ok and we will not "leak" this GraphStage because of not calling `completeStage`. I debugged it and seemed like `postStop` is always called (no matter if upstream and downstream just completed or failed) but I lack knowledge to be completely sure.